Ziko's "Guess Who" Game

vrill opened the pull request:
This was fun, and very challenging. Still some things I can't really wrap my head around, but we're just getting started... :) Eager for feedback!

vrill committed:
Initial commit of the Guess Who game

vrill committed:
Optimizations, final commit (?)

Added a title, fixed so drop-down is reset on restart, and optimized a
few handles for clarity.

vrill committed:
Changed options value 'hats' for grammar purposes

Easier than to re-write the whole JS logic for the 'other' category when
the guess is made and prompt displayed for the user, as we're using booleans
rather than strings for that category, at least for my skill level.
Though 'facehair' isn't really proper English.
Compatible use if in Swedish, however, so I'll just let that be for now... :)

vrill committed:
Changed the select value for hat and facehair

Final change for grammar purposes. Updated the array keys to properly read the
re-named values.

vrill committed:
Readme filled out

What is it, what the problem was; how it was solved, and where to view
it. Enjoy.

vrill committed:
Final clean-up

vrill committed:
New effect

Added a hover effect on the cards, for smoother transition. CSS isn't my
forte, but it at least 'feels' better playing the game now. IMHO. :)

HIPPIEKICK commented on code/script.js:
```diff
@@ -11,80 +15,98 @@ const CHARACTERS = [
     eyeColor: 'hidden',
     glasses: true,
-    hat: true,
+    'a hat': true,
     smoker: false,
+    'face hair': false,
+    jewelry: false,
```
Nice additions 👍

HIPPIEKICK replied:
Minor thing, but the last property in an object doesn't need a comma after 😇

HIPPIEKICK commented on code/script.js:
```diff
@@ -244,100 +296,155 @@ const generateBoard = () => {
-  // What else should happen when we start the game?
+  board.style.display = 'flex'
+  winOrLose.style.display = 'none'
+  questions.selectedIndex = '0'
```
I'm not sure what this is doing here, maybe you wanted to implement a counter that keeps track of the number of guesses? In that case, you could declare a global variable and update the value of it every time the checkMyGuess is invoked or similar :)

vrill replied:
`board.style.display = 'flex'` makes sure the `board` section is being displayed upon start, basically reverses what the `checkMyGuess`  function does in order to display the `winOrLose` section.
`winOrLose.style.display = 'none'` sets the `winOrLose` section back to its original state of `none`.

Whereas `questions.selectedIndex = '0'` just resets the drop-down selection to the first option (`disabled selected`), instead of displaying the last guess in the drop-down.

I tried various `location.reload()`-esque solutions, but neither worked as I wanted them to... At my current level this was what I came up with. :) No plan for any counter, but that could be a cool addition to track the score or how many guesses were made on average to reach the correct answer!

HIPPIEKICK replied:
Ahh of course 🤦‍♀️  You sure have attention to detail! 💯

HIPPIEKICK commented on code/script.js:
```diff
@@ -244,100 +296,155 @@ const generateBoard = () => {
-  // We also need a variable that stores the actual value of the question we've selected.
-
-  if (category === 'hair color') {
+  const guessCategory = questions.options[questions.selectedIndex].parentNode.label
+  const guessValue = questions.value
+  // console.log(guessValue)
+  // console.log(guessCategory)
```
Good naming! You can remove console.logs in production :)

HIPPIEKICK commented on code/script.js:
```diff
@@ -244,100 +296,155 @@ const generateBoard = () => {
-  // Compare the currentQuestion with the secret person.
-  // See if we should keep or remove people based on that
-  // Then invoke filterCharacters
+  const guess = currentQuestion.value === secret[currentQuestion.attribute]
+  // console.log(guess)
+  filterCharacters(guess)
```
I like that you are renaming the variables! Since this is a boolean, it could be named something like `isGuessCorrect` or something to indicate that its value is either true or false

HIPPIEKICK commented on code/script.js:
```diff
@@ -244,100 +296,155 @@ const generateBoard = () => {
+        }
+    } else if (category === 'other') {
+      if (guess) {
+        alert(`✔ Yes, ${attribute} is the name of the game! Good guess.
+🤖 Keeping those ones.`)
+      } else {
+        alert(`❌ Nope, ${attribute} isn't in the cards! Guess again.
+🤖 Removing those ones.`)
```
Clever ;)

HIPPIEKICK commented on code/script.js:
```diff
@@ -244,100 +296,155 @@ const generateBoard = () => {
-  // store the interaction from the player in a variable.
-  // remember the confirm() ?
-  // If the player wants to guess, invoke the checkMyGuess function.
+  confirm(`🙈 You're about to guess on ${suspect}. Are you sure?`)
+  if (confirm) {
+    checkMyGuess(suspect)
```
If you don't confirm, it will still make the guess 😬 Try saving the confirm in a variable and use that, something like:
```
const makeAGuess = confirm(`🙈 You're about to guess on ${suspect}. Are you sure?`)
  if (makeAGuess) {
    checkMyGuess(suspect)
  }
```

vrill replied:
Excellent, I didn't QA this properly I see... This has been fixed & pushed. Thank you! :D

HIPPIEKICK commented on code/script.js:
```diff
@@ -244,100 +296,155 @@ const generateBoard = () => {
+
+// If the user selects a value from the drop-down,
+// pass its value to the selectQuestion function
+questions.addEventListener('change', () => selectQuestion())
```
I realize this is how we wrote it in the Q/A as well, but since we no longer pass along an argument, we can write this like the others:
`questions.addEventListener('change', selectQuestion)`

HIPPIEKICK commented on code/style.css:
```diff
@@ -71,37 +72,43 @@ select {
   flex-direction: column;
   align-items: center;
   justify-content: space-between;
+  transition: 0.3s;
```
Nice with the transitions 👍

HIPPIEKICK reviewed (commented):
Good job with this project Christoffer! I like that you made it your own by adding more properties, changing colors, and coming up with good variable names (that's one of the hardest things!).

vrill committed:
Version 1.1

Minor clean up and a fix after proper feedback. Cheers!

change: remove unneccesary comma after last property in the array
change: rename function to be more in-line with its value; boolean
change: shorten function length, since it doesn't pass any argument
change: remove console.log entries
fix: guess confirmation didn't honor a decline
change: guess confirmation function renamed for added clarity
